### PR TITLE
feat(server): route metadata

### PR DESCRIPTION
**Pull request by LouisHaftmann**
Implementation of route metadata #1731

### CODE REVIEW

File: packages/server/src/internals/procedure.ts [hunk 1]
```diff
@@ -44,11 +44,12 @@ export type ProcedureResolver<TContext, TInput, TOutput> = (opts: {
   type: ProcedureType;
 }) => Promise<TOutput> | TOutput;
 
-interface ProcedureOptions<TContext, TInput, TOutput, TParsedOutput> {
-  middlewares: Array<MiddlewareFunction<any, any>>;
+interface ProcedureOptions<TContext, TInput, TOutput, TMeta, TParsedOutput> {
```
Comment (KATT) [inline packages/server/src/internals/procedure.ts:47]: Align arg order to `Router`

```suggestion
interface ProcedureOptions<TContext, TMeta, TInput, TOutput, TParsedOutput> {
```

File: packages/server/src/router.ts [hunk 2]
```diff
@@ -528,11 +571,14 @@ export class Router<
    * Merge router with other router
    * @param router
    */
-  public merge<TChildRouter extends Router<TContext, any, any, any, any, any>>(
+  public merge<
+    TChildRouter extends Router<TContext, any, any, any, any, any, any>,
```
Comment (KATT) [inline packages/server/src/router.ts:575]: We don't wanna merge routers that have different meta shapes (at least for now)


```suggestion
    TChildRouter extends Router<TContext, any, TMeta, any, any, any, any>,
```

File: packages/server/src/router.ts [hunk 3]
```diff
@@ -528,11 +571,14 @@ export class Router<
...
+  >(
     router: TChildRouter,
   ): Router<
     TInputContext,
     inferRouterContext<TChildRouter>,
+    TMeta & inferRouterMeta<TChildRouter>,
```
Comment (KATT) [inline packages/server/src/router.ts:581]:
```suggestion
    TMeta,
```

File: packages/server/src/router.ts [hunk 4]
```diff
@@ -546,13 +592,14 @@ export class Router<
    */
   public merge<
     TPath extends string,
-    TChildRouter extends Router<TContext, any, any, any, any, any>,
+    TChildRouter extends Router<TContext, any, any, any, any, any, any>,
```
Comment (KATT) [inline packages/server/src/router.ts:595]:
```suggestion
    TChildRouter extends Router<TContext, any, TMeta, any, any, any, any>,
```

**Review by KATT — DISMISSED** [review]
This looks great! I'm impressed that you were able to dive deep and get this in place.

Left comments here. Please resolve and dismiss my current review so I get a notification :)

Furthermore, this change needs tests & docs updates, maybe add a new `router-meta.test.ts`-file where it's tested.

**Review by KATT — APPROVED** [review]
@all-contributors add @LouisHaftmann  for code,tests,docs

**Comment by KATT** [comment]
🔥 AMAZING

**Review by sachinraja — APPROVED** [review]
Love this! I think this can help a lot with building third-party extensions and with a few ideas I have. Implementation wise it looks good too (wow that's a lot of changes for one generic). 

As a side note I think the other benefits of this will also be addressed by the reusable procedures in the v10 playground.

**Review by jlalmes — APPROVED** [review]
Brilliant work, thank you @LouisHaftmann! 🚀 

This is going to be super useful for greater control of auto-generated OpenAPI specs moving forward (https://github.com/trpc/trpc/issues/1724) 

Let's merge ✅

**Comment by jlalmes** [comment]
I think there's settings to only allow squash commits from PR's.
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests

**Comment by KATT** [comment]
> I think there's settings to only allow squash commits from PR's. https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests

Yeah, but I need the merge ones for the `next`-branch and you can't decide it on a per-branch basis.

**Comment by KATT** [comment]
@LouisHaftmann could you re-open this PR? I wanna merge it with a squash https://github.com/trpc/trpc/compare/main...feat/route-metadata?expand=1

**Comment by LouisHaftmann** [comment]
> Oh man, this messed with the pretty commit history of `main` so I will do some git tricks to fix it

sry sry sry

**Pull request merged**